### PR TITLE
REF: Support Extract trait refactoring for traits

### PR DESCRIPTION
**Pull request by dima74**
Meta issue: #8326

Please see #8132 for general overview of the refactoring

changelog: Support `Extract trait` refactoring for traits

### CODE REVIEW

**Review by ortem — APPROVED** [review]
LGTM

File: src/main/kotlin/org/rust/ide/refactoring/extractTrait/RsExtractTraitHandler.kt [hunk 1]
```diff
@@ -38,15 +40,16 @@ class RsExtractTraitHandler : RefactoringActionHandler {
     override fun invoke(project: Project, editor: Editor, file: PsiFile, dataContext: DataContext?) {
         val offset = editor.caretModel.offset
         val element = file.findElementAt(offset) ?: return
-        val impl = element.ancestorOrSelf<RsImplItem>() ?: return
-        if (impl.traitRef != null) return
-        if (!CommonRefactoringUtil.checkReadOnlyStatus(project, impl)) return
+        val traitOrImpl = element.ancestorOrSelf<RsTraitOrImpl>() ?: return
+        if (traitOrImpl is RsImplItem && traitOrImpl.traitRef != null) return
```
Comment (ortem) [inline src/main/kotlin/org/rust/ide/refactoring/extractTrait/RsExtractTraitHandler.kt:44]: [minor] `if` checks could be replaced with `when` to emphasize that it is either `RsImplItem` or `RsTraitItem`, like `org.rust.lang.core.types.ExtensionsKt#getSelfType` for example

Comment (dima74) [reply src/main/kotlin/org/rust/ide/refactoring/extractTrait/RsExtractTraitHandler.kt:44]: We have second condition and `return` here, so I am not sure how best to do it

**Comment by dima74** [comment]
bors r=ortem

**Pull request merged**